pkg/backup/restore_job.go

```diff
@@ -1816,6 +1816,21 @@ func (r *restoreResumer) doResume(ctx context.Context, execCtx interface{}) erro
 		return err
 	}
 
+	if details.OnlineImpl() && len(details.DownloadSpans) == 0 {
+		// Persist the download spans before the link phase begins as OnFailOrCancel
+		// could use them if called.
+		downloadSpans, err := getDownloadSpans(p.ExecCfg().Codec, preData, mainData)
+		if err != nil {
+			return err
+		}
+		// Ensure we have the latest copy of the job details.
+		details = r.job.Details().(jobspb.RestoreDetails)
+		details.DownloadSpans = downloadSpans
+		if err := r.job.NoTxn().SetDetails(ctx, details); err != nil {
```

Comment:
Are there any concurrent processes that might be writing to details after the job starts? I'm wondering if we have to worry about overwriting a some change that committed between when the job started and when we checkpoint the details.

Reply:
I am not, as the restore flow has not begun yet. While it is sad that we have to update job details after job planning, this felt like the change with the lowest blast radius. In the func above (`createImportingDescriptors`) we also update the job details. 

I think this is safe so long as a refresh my local copy of `details`. All updates to details will also update the in memory copy in `r.job.Details()`

Reply:
`createImportingDescriptors` is using a txn, but you are right that it doesn't actually read the details in the transaction. So at the very least this isn't introducing a new bug. 

In general I would prefer if we read the details, update, and write them in a transaction so that we don't need to think carefully about what code paths might be updating details. But if this is common practice for jobs 🤷 As long as we are consistent with the convention and only the job is allowed to update the details outside of a txn.

Reply:
big yak -> refactoring how jobs update their state